src/LanguageServer/Protocol/Workspaces/LspMiscellaneousFilesWorkspaceProvider.cs

```diff
@@ -3,27 +3,130 @@
+    /// 
+    /// Calls to this method and <see cref="TryRemoveMiscellaneousDocument(Uri, bool)"/> are made
+    /// from LSP text sync request handling which do not run concurrently.
+    /// </summary>
+    public TextDocument? AddMiscellaneousDocument(Uri uri, SourceText documentText, string languageId, ILspLogger logger)
+    {
+        var documentFilePath = ProtocolConversions.GetDocumentFilePathFromUri(uri);
+
+        var container = new StaticSourceTextContainer(documentText);
+        if (metadataAsSourceFileService.TryAddDocumentToWorkspace(documentFilePath, container, out var documentId))
+        {
+            var metadataWorkspace = metadataAsSourceFileService.TryGetWorkspace();
+            Contract.ThrowIfNull(metadataWorkspace);
+            var document = metadataWorkspace.CurrentSolution.GetRequiredDocument(documentId);
+            return document;
+        }
+
+        var languageInfoProvider = lspServices.GetRequiredService<ILanguageInfoProvider>();
+        if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+        {
+            // Only log here since throwing here could take down the LSP server.
+            logger.LogError($"Could not find language information for {uri} with absolute path {documentFilePath}");
+            return null;
+        }
+
+        var sourceTextLoader = new SourceTextLoader(documentText, documentFilePath);
+
+        var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument(
+            this, documentFilePath, sourceTextLoader, languageInformation, documentText.ChecksumAlgorithm, Services.SolutionServices, []);
+        OnProjectAdded(projectInfo);
+
+        if (languageInformation.LanguageName == "Razor")
+        {
+            var docId = projectInfo.AdditionalDocuments.Single().Id;
+            return CurrentSolution.GetRequiredAdditionalDocument(docId);
+        }
+
+        var id = projectInfo.Documents.Single().Id;
+        return CurrentSolution.GetRequiredDocument(id);
+    }
+
+    /// <summary>
+    /// Removes a document with the matching file path from this workspace.
+    /// 
+    /// Calls to this method and <see cref="AddMiscellaneousDocument(Uri, SourceText, string, ILspLogger)"/> are made
+    /// from LSP text sync request handling which do not run concurrently.
+    /// </summary>
+    public void TryRemoveMiscellaneousDocument(Uri uri, bool removeFromMetadataWorkspace)
+    {
+        var documentFilePath = ProtocolConversions.GetDocumentFilePathFromUri(uri);
+        if (removeFromMetadataWorkspace && metadataAsSourceFileService.TryRemoveDocumentFromWorkspace(documentFilePath))
+        {
+            return;
+        }
+
+        // We'll only ever have a single document matching this URI in the misc solution.
+        var matchingDocument = CurrentSolution.GetDocumentIds(uri).SingleOrDefault();
+        if (matchingDocument != null)
```

Comment:
curious when this is null.  i don't feel like i ahve a good sense of the invariants we have in this space.  feels loosey goosey.